File: main_project/tripod-new-branch/src/java/TripodConfig/src/main/java/com/renren/tripod/configutil/ConfigData.java
package com.renren.tripod.configutil;

import java.util.HashMap;
import java.util.Map;
import java.util.Map.Entry;

public class ConfigData implements Cloneable {

	public static String TRIPOD_MANAGER = "Manager";
	public static String TRIPOD_PRODUCER = "Producer";
	public static String TRIPOD_CACHE = "Cache";
	public static String TRIPOD_PRODUCER_CC = "cc";
	public static String TRIPOD_PRODUCER_JAVA = "java";

	public String _name = "";
	public String _value = "";
	public String _version = "";

	public Map<String, ConfigData> _childrenDatas = new HashMap<String, ConfigData>();

	public String toString() {
		StringBuffer buff = new StringBuffer();
		buff.append("\nNAME: " + _name);
		buff.append("\nVALUE: " + _value);
		buff.append("\nVERSION: " + _version);
		for (Entry<String, ConfigData> cd : _childrenDatas.entrySet()) {
			buff.append("   " + cd.getKey().toString() + "\n");
		}
		return buff.toString();
	}

	public ConfigData clone() {
		ConfigData newData = new ConfigData();
		newData._name = _name;
		newData._value = _value;
		newData._version = _version;
		for (Entry<String, ConfigData> ent : _childrenDatas.entrySet()) {
			newData._childrenDatas.put(ent.getKey(), ent.getValue().clone());
		}
		return newData;
	}
}

File: main_project/tripod-new-branch/src/cpp/configutil/ConfigData.cpp
#include "ConfigData.h"

namespace com {
namespace renren {
namespace tripod {

ConfigData::ConfigData() {
    name_ == "";
    value_ == "";
    version_ == "";
}

ConfigDataPtr ConfigData::getData() {
    ConfigDataPtr newData(new ConfigData);
    newData->name_ = name_;
    newData->value_ = value_;
    newData->version_ = version_;

    for(std::map<std::string, ConfigDataPtr>::const_iterator it = childrenDatas_.begin(); it != childrenDatas_.end(); ++it) {
        newData->childrenDatas_.insert(make_pair(it->first, it->second->getData()));
    }

    return newData;
}

std::string ConfigData::str() {
    std::ostringstream buff;
    buff << "ConfigData [";
    buff << "name=" << name_ << ";";
    buff << "value=" << value_ << ";";
    buff << "version=" << version_ << ";";
    buff << "children size=" << childrenDatas_.size() << ";";

    for (std::map<std::string, ConfigDataPtr>::iterator it = childrenDatas_.begin(); it != childrenDatas_.end(); ++it) {
        buff << "child data [" << it->second->str();
    }

    buff << " ]";

    return buff.str();
}

};
};
};
